### ISSUE TITLE
Validation errors after command encoder finishes

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#2264]
It was agreed at some point that validation errors in commands wouldn't be exposed immediately through the current error scope, but instead would just change the state of the encoder and not be exposed to the error scope until `finish()`.

However once the command encoder has already been finished, what should happen when a command is attempted anyway?

1. Silently fail.
2. Raise an error to the error scope even though encoding commands don't usually do that. (Dawn does this.)
3. Change encoding commands to normally raise errors to the error scope.
4. [EDIT:] Raise an exception if the content-side state is "finished" (or 4b: if it's wrong at all). Currently the spec says "encoder state" (open, render pass, compute pass, or closed) is a content-timeline state, actually, so spec-wise this is easy.

**Comment by kainino0x** [comment]
Unfortunately I don't remember the exact reasoning behind making commands work this way. I seem to remember it was a performance consideration, so does option 2 defeat the purpose? (It seems fine to me offhand, but so does option 3.)

One thing I've imagined is that command encoding could be entirely content-timeline (with the encoded result being sent all at once and validated together). Implementing option 2 in this case would require the content timeline to conditionally inject an error into the error scope. Fortunately, the state needed for validation is simple, so validation could be done in the content timeline.

---

Option 2 would make commands look like this:
- if encoder is closed: raise validation error to error scope
- if encoder is otherwise not in the right state or command validation fails: put encoder in "failed" state
- encode command

**Comment by kainino0x** [comment]
Added option 4 because I realized the spec already specifies the encoder state as content-side.

**Comment by Kangz** [comment]
Option 4 looks nice, it's an extra condition in pretty performance critical code, but it's probably ok.

**Comment by kainino0x** [comment]
Talked with @toji and after a lot of indecision we're _kind of_ leaning toward option 3. Removing the special-ness of encoder commands removes the ambiguity over how this should work. It also benefits developers because they can possibly see multiple errors from the same encoder instead of only seeing the first one each time.

Option 4 is appealing because it makes some trivial validation synchronous, which is much easier for developers - but we've been intentionally designing the rest of the API without considering that, because we would prefer to have a dev tools "break on webgpu error" mode instead.

Option 2 raises errors at a reasonable point (asynchronously), without doing so in an inconsistent style like option 4.

Option 3 still does that, minus the current special handling.

**Comment by Kangz** [comment]
I think the issue with 2) is that some implementation like Mozilla's would like to keep encoders completely in the renderer process until `finish()`. This makes it hard to integrate with error scopes because the timelines don't line up.

**Comment by kainino0x** [comment]
I think you meant option 3? If recording is done in the content process, then option 2 just requires "if already finished, inject error", while option 3 requires (duplicated) content-side validation to determine when to inject errors.

**Comment by Kangz** [comment]
Yep I meant 3).

**Comment by kainino0x** [comment]
Editors discussion:
- 4 would require us to _also_ implement 1 or 2 in browsers because the gpu process can't trust validation done by the client. Also inconsistent with other errors.
- 1 is unfriendly.
- 3 limits the range of possible implementations (see ^), and would require refactoring work in both wgpu and dawn since they both take advantage of deferred errors.
- 2 is easy enough to implement even with content-side validation, only requires a way to "injectError" 

So I think we will stick with option 2. Needs some minor specification work (after #2284).

**Comment by kainino0x** [comment]
(Moving directly to Needs Specification because it doesn't conflict with what the group previously decided about encoder error handling.)

**Comment by kainino0x** [comment]
While working on #2312 I found I would like to extend this to passes as well: once a pass has ended, attempting to encode commands generates a validation error instead of marking the encoder as invalid.

This is nice because it means invalid commands can just "locally" make their pass encoder invalid instead of needing a way to reach out to the top-level encoder and make that invalid after endPass() has already happened.
